src/cached_object_store.rs

```diff
@@ -762,6 +798,214 @@ impl LocalCacheEntry for FsCacheEntry {
+        }
+    }
+
+    pub async fn maybe_evict(&self, bytes: u64) {
+        if self.limit_bytes == 0 {
+            return;
+        }
+
+        self.tx.send(bytes).await.ok();
+    }
+}
+
+struct FsCacheEvictorInner {
+    root_folder: std::path::PathBuf,
+    tracked_bytes: AtomicU64,
+    limit_bytes: u64,
+    evict_buffer: Mutex<VecDeque<std::path::PathBuf>>,
+}
+
+impl FsCacheEvictorInner {
+    pub async fn maybe_evict(&self, bytes: u64) {
+        self.tracked_bytes.fetch_add(bytes, Ordering::SeqCst);
+        if self.tracked_bytes.load(Ordering::Relaxed) <= self.limit_bytes {
+            return;
+        }
+
+        let evict_factor = 10;
+        for _ in 0..evict_factor {
+            let evicted_bytes = self.evict_once().await;
+            if evicted_bytes == 0 {
+                return;
+            }
+            self.tracked_bytes
+                .fetch_sub(evicted_bytes, Ordering::SeqCst);
+        }
+    }
+
+    // evict once, return the bytes evicted
+    async fn evict_once(&self) -> u64 {
+        let (target, target_bytes) = match self.pick_evict_target().await {
+            Some(target) => target,
+            None => return 0,
+        };
+
+        match tokio::fs::remove_file(&target).await {
+            Err(err) => {
+                warn!("evictor: failed to remove the cache file: {}", err);
+                0
+            }
+            Ok(()) => {
+                debug!(
+                    "evictor: evicted cache file: {:?}, bytes: {}",
+                    target, target_bytes
+                );
+                target_bytes
+            }
+        }
+    }
+
+    pub async fn scan_cache_folder_bytes(
+        root_folder: &std::path::PathBuf,
+    ) -> Result<usize, Box<dyn std::error::Error>> {
+        let mut total_bytes = 0;
+        let iter = WalkDir::new(root_folder).into_iter();
+        for entry in iter {
+            let entry = entry?;
+            let metadata = tokio::fs::metadata(entry.path()).await?;
+            total_bytes += metadata.len();
+        }
+        Ok(total_bytes as usize)
+    }
+
+    // pick a file to evict, return None if no file is picked. it takes a pick-of-2 strategy, randomized pick
+    // two of files, compare their last access time, and evict the older one.
+    async fn pick_evict_target(&self) -> Option<(std::path::PathBuf, u64)> {
+        // extend the evict_buffer if it's empty. it will randomized iterate the cache folder,
+        // and randomly pick 1000 files into the evict_buffer.
+        let mut rng = rand::rngs::StdRng::from_entropy();
+        if self.evict_buffer.lock().await.len() < 2 {
+            let iter = WalkDir::new(&self.root_folder).into_iter();
+            for entry in iter.choose_multiple(&mut rng, 1000) {
+                let entry = match entry {
+                    Ok(entry) => entry,
+                    Err(err) => {
+                        warn!("evictor: failed to walk the cache folder: {}", err);
```

Comment:
nit:
this warning message can be misleading, I think. is it always the case that this error branch will only be reached if the very first `entry`  unpacks into `Err(err)`. Otherwise, if this execution branch can be reached after the 666th randomly chose file/entry, then it wouldn't be accurate to tag this as a `failure to walk the cache folder` (which reads more like a showstopper for the intended goal of this function). Instead, this should be considered a partial failure that should be tagged with a different warning message.